include/rapidcheck/detail/ApplyTuple.h

```diff
@@ -13,7 +13,7 @@ struct ApplyTupleImpl;
 
 template <typename... Ts, std::size_t... Indexes, typename Callable>
 struct ApplyTupleImpl<std::tuple<Ts...>, Callable, IndexSequence<Indexes...>> {
-  using ReturnType = typename std::result_of<Callable(Ts &&...)>::type;
+  using ReturnType = typename std::result_of<Callable&&(Ts &&...)>::type;
```

Comment:
I'm afraid I'm not following the reasoning for making this an rvalue here. It's not called as an rvalue below so I don't see how this is correct?

Reply:
`std::result_of<F(Args...)>::type` does not work if F is a function (but works if F is a reference to function).
Here's an example:
```
int foo(int x) { return 0; }
using T = std::result_of<decltype (foo)(int)>::type;
```
^ fails with 
```
./foo.cpp:34:39: error: function cannot return function type 'decltype(foo)' (aka 'int (int)')
```
...however, this works:
```
using T = std::result_of<decltype (&foo)(int)>::type;
```

When I do:
```
template <typename T, typename... Args>
Gen<T> construct(Gen<Args>... gens) {
  return gen::map(gen::tuple(std::move(gens)...),
                  [](std::tuple<Args...> &&argsTuple) {
                    return rc::detail::applyTuple(
                        std::move(argsTuple),
                        [](Args &&... args) { return T{std::move(args)...}; });
                      std::forward<std::tuple<Args...>>(argsTuple),
                      detail::construct<T, Args...>
                    );
                  });
}
```
...it fails when it tries to do `result_of<decltype(detail::construct<T, Args...>)(Args...)`, while I want it to do `result_of<decltype(&detail::construct<T, Args...>)(Args...)`.

https://en.cppreference.com/w/cpp/types/result_of#Notes suggests using `result_of<F&&(Args...)>` instead `result_of<F(Args...)>` to avoid this.

Reply:
Because of the limitations of `result_of` I think it is better to keep the `&&`. Technically speaking this is not a r-value reference, but does reference collapsing based on the value category of `Callable` so there should be no substantial changes.